### ISSUE TITLE
RFC: BLE support

### ISSUE THREAD

**Issue by igrr** [espressif/arduino-esp32#423]
This ticket is opened to discuss possible approaches to implementing BLE wrappers, desired API and implementation choices.

_**Note**: please don't use this issue to express your desire/need to have BLE support in the form of "+1" comments, questions on schedule and such. Instead, please use "Subscribe" button on the right to receive email updates on the discussion._

Prior art
--------
- [CurieBLE](https://github.com/01org/corelibs-arduino101/tree/master/libraries/CurieBLE/src) for Arduino 101
- [sandeepmistry/arduino-BLEPeripheral](https://github.com/sandeepmistry/arduino-BLEPeripheral/tree/master/src)
- [Mbed BLE library](https://github.com/RedBearLab/nRF51822-Arduino/tree/S130/arduino-1.6.x/hardware/RBL/RBL_nRF51822/cores/RBL_nRF51822/ble-master/ble) (as seen in RedBearLab/nRF51822-Arduino)
- not very Arduino-esque [NordicSemiconductor/ble-sdk-arduino](https://github.com/NordicSemiconductor/ble-sdk-arduino)
- something else? Add a comment if you know of other libraries.

Support for ESP32 BLE in other frameworks
-------------------------------------------
- [micropython](https://github.com/micropython/micropython-esp32/pull/73) — BLE implementation by Pycom
- [micropython](https://github.com/micropython/micropython-esp32/pull/86) — new BLE implementation
- [nodemcu](https://github.com/nodemcu/nodemcu-firmware/blob/dev-esp32/components/modules/bthci.c) — HCI layer only

Resources
----------
- [Doxygen-generated documentation](https://docs.espressif.com/idf/en/latest/api-reference/bluetooth/index.html) for ESP-IDF Bluetooth stack
- ESP-IDF Bluetooth [examples](https://github.com/espressif/esp-idf/tree/master/examples/bluetooth)

Action list
----------
- [ ] Decide which existing library to choose as a base for the API
- [ ] Design of advertising API
- [ ] Design of Peripheral/server API
- [ ] Design of Central/client API
- [ ] Implementation of advertising API
- [ ] Implementation of Peripheral/server API
- [ ] Implementation of Central/client API

Everyone is welcome to offer suggestions and comments about API design. Contributing actual code is also a great way to move this issue forward; however please post the proposed API for comment before starting implementation.

**Comment by igrr** [comment]
Having taken a look at the existing libraries and their example code, i like the way sandeepmistry/arduino-BLEPeripheral is designed. The chip-specific part is separated reasonably well to attempt shoehorning ESP32 support into that library. The way adding services and characteristics works in this library will probably play nice with GATT table APIs in the ESP-IDF.
Would be nice if someone else would have a brief look at the code and examples to see if there are any things which will make porting problematic.

**Comment by WALLTECH** [comment]
FWIW, having worked with the Curie BLE libraries they are a pleasure to use and vote for that. Obvious preference to under the hood ease of conversion and time of course.

**Comment by me-no-dev** [comment]
Probably the only major missing piece for me (in IDF) is the ability to properly pair with a device and hold that key, so the same device can reconnect again without peering. The rest I somewhat understand and have working. Have not tried to be gatt client and server at the same time also. I wonder if that is possible or we need to figure out a safeguard to prevent it. Properly connecting an item in the gatt server table to id so you can alter that item is also not so clear. Using human logic I can get to it, but that could not work in code ;)

**Comment by zencuke** [comment]
Thank you. "Other frameworks:" I've been prototyping my central mode app in Xcode/OSX mostly because I know Xcode environment for command line apps and Apple's Core BLE library is well documented. There are lots of example projects. That works as a self education process but a proprietary closed library example is obviously less helpful for this project. I mention it for completeness. It is a working ble library.

**Comment by kriswiner** [comment]
We're using Nordic's SDK (not well) to enable both central and peripheral roles for custom nRF52 devices. I think BLEPeripheral cannot accomodate central role specification at the moment. I would just say that this is a serious deficiency and I would encourage the BLE API developers to take the need for both central and peripheral roles into account in the design of the ESP32 BLE API. 

I do like the syntax BLEPeripheral has chosen to specify services and characteristics and this would be a good choice for the ESP32 to emulate. I have found this much easier to use than mbed BLE.

**Comment by skateone** [comment]
i am also trying to get the esp32 to communicate with a Nordic nRF52 device. in my case we try setup the esp32 with a GATTs server hosting services and characteristics and try use the Nordic to discover the ESP32. for some reason we can only see the device from the ESP32 and non of the services.

**Comment by S3ler** [comment]
I started thinking of a [Bluetooth Central API Implementation](https://github.com/sandeepmistry/arduino-nRF5/issues/166#issuecomment-309287113) for the nRF5 Chips. It is a good Idea to use overall the same API so i participate here.
I agree with [sandeepmistry to use the CurieBLE API](https://github.com/sandeepmistry/arduino-nRF5/issues/166#issuecomment-309281588). It provides both API for Central, Peripheral and is already used in an Arduino project.

**Comment by copercini** [comment]
I really love @nkolban C++ implementation:
https://www.youtube.com/watch?v=WGozJXuQ8dY&feature=youtu.be
https://github.com/nkolban/esp32-snippets/tree/master/cpp_utils

It's clean and already works with ESP32, so just need a minimal effort make an Arduino port

**Comment by ouki-wang** [comment]
Do you decide which api to use? I vote for curieBLE.

**Comment by jcmojj** [comment]
Do vou know if  de can check signal strenght with it? It is all I Will
Need..

**Comment by ayavilevich** [comment]
Sandeep's library is very easy to work with and a nicely designed API, but as mentioned above it doesn't support Central roles or running Central at the same time as running Peripheral. I would like to see a library that is as easy to understand as Sandeep's but can support Central with many Peripherals and Central and Peripheral roles at the same time.

Naturally, it makes sense to get an initial version out quickly, even if it is only Peripheral. Thanks.

**Comment by d3cline** [comment]
CurieBLE is my personal vote as far as an API to emulate, Intel has been pushing the Arduino 101 and the curie chip a lot recently, giving away a bunch on hackster (i got one) The API is nice once you understand it.

**Comment by WALLTECH** [comment]
I agree, and it's ok to update the library as features become available rather than holding off until it's done. Different people need different features and as they come out people should be able to use them.

**Comment by gesaleh** [comment]
I agree for the Sandeep's library easy to use, even if i had some issue with RB blend 2, 
if there is any progress I would like to test it. for now I need only the Peripheral (sending sensors from ESP to Pi) that would be a great start

**Comment by tijnkooijmans** [comment]
+1 for CurieBLE.

**Comment by tijnkooijmans** [comment]
I also like the CurieBLE API.

**Comment by scramble45** [comment]
CurieBLE API +1

**Comment by copercini** [comment]
So, Intel has announced the end-of-life timeline for the Intel Curie module, and will discontinue manufacturing Arduino 101: https://communities.intel.com/thread/116434

**Comment by WALLTECH** [comment]
@copercini this doesn't change the simplicity and suitability of the CurieBLE library. The hardware cost and misjudging the market for hardware needs was what's closing it down.

**Comment by MarosValter** [comment]
I vote for CurieBLE

**Comment by nkolban** [comment]
Howdy folks,
I have been tinkering with a BLE API in C++ on the ESP32 that exposes an API that I came up with ... see:

https://esp32.com/viewtopic.php?f=13&t=2330

Is anyone actively working on a different/alternate encapsulation of ESP32 BLE support?

I see discussion of desire for a BLE API that matches that provided as described here:

https://www.arduino.cc/en/Reference/CurieBLE

Is there a designated sub-project leader for the technical discussion described in this Github issue?  What would be a shame is if we all went in separate directions and perform individual work as opposed to a collaborative effort.  I'll be delighted to pitch in in some way.  It may be as simple as taking my existing library of BLE C++ functions are re-exposing the interfaces to match those of CurieBLE or maybe there is a design proposal that starts from scratch?

**Comment by d3cline** [comment]
I am waiting to see what the official decision is by espreiff and igrr. I wish I could provide more feedback from the hardware end. I need a little boiler plate to run with and I am willing to help debug and provide more feedback. I am anxious to start coding with [Kotlin](https://github.com/objectsyndicate/Kotlin-BluetoothLeGatt), BLE, and ESP32. The [arduino](https://github.com/objectsyndicate/ioTank-101/blob/master/ioTank101.ino#L222) code I already have which was for Arduino-101 and what will be similar in future projects just needs to send a JSON string. Basically waiting to see what will happen before I see what/if any assistance I can lend beyond debugging.

**Comment by Wekhrozomedo** [comment]
CurieBLE API

**Comment by gmag11** [comment]
I'm following tutorials from @nkolban and trying to replicate same code on Arduino. The problem is that included precompiled IDF is behind main esp32-idf repository, and many of the functions there are not implemented here.

Anyway I find his implementation simple enough (on the interface) and close to esp32 capabilities.

I'd vote to see his work integrated on arduino-esp32.

**Comment by jcmojj** [comment]
So far I see that CurieBle was the most voted. I vote against it because
for me it is very important to read signal strength  (rssi) and CurieBle as
far as I read is unable to do it. The kolban could do it.

**Comment by dineshjoshi** [comment]
I like CurieBLE as well.

**Comment by WALLTECH** [comment]
@jcmojj Of course CurieBLE can do signal strength. Use .rssi() on a BLEDevice object to get it.

int BLEDevice::rssi() const
{
    return BLEDeviceManager::instance()->rssi(this);
}

**Comment by d3cline** [comment]
Do we have any ETA? The market doesn't wait. Reengineering low level APIs isn't my focus, but picking what is ready and building products is. I thought the ESP32 was more developed and supported, BLE has been discussed for about a year now, there is a 'simple example' which shouldn't even exist since it's incomplete and misleading. How much longer are we going to have to wait for even an alpha/beta or some kind of decision? What is the current decision by the upstream providers? Are they waiting for the community to write something? The intel curie chip still can be bought without an arduino, I can still run that code in my products. I could use the ESPIDF if i wanted to port my code to a low level, proprietary format, and spend hours learning how BLE works at a hardware level, but I won't do that as I am only 1 person and have so much effort to place elsewhere. Arduino matters because rapid development matters, BLE is the 1 feature we need from ESP32 that we can't get from the 8266, why is it not a major priority and done already?

**Comment by zencuke** [comment]
@d3cline. I  feel your pain but your post was a bit undirected. You may have posted  it in the wrong place. Do you understand the purpose of this particular forum and who is likely to read it? Maybe you could clarify your post. Exactly who is it that you think has the wrong priorities? And who do you think should be answering your questions and addressing the issues you raise? Who do you think should have already solved them?

**Comment by clowrey** [comment]
@d3cline Curie module is EOL (end of life) according to an email I got from Digikey a few days ago. So not recommended for new designs.

**Comment by d3cline** [comment]
Espriff is the upstream provider, and they have paid developers, one of them started the thread, they have dumped a ton of time into the IDF. They are in the best position to make this move forward. It is in their interest to as well since many of us are waiting to put the ESP32 in use, but if I have to decide on another MCU since it will take another few months I will.

**Comment by d3cline** [comment]
Or let me put it this way, I would order a few hundred right now from aliexpress if I knew this was working and I could check it off my logistics checklist for my next project. Since it may be 'whenever', the thread is 2 months old, and BLE demand in general is almost a year old, I am looking into alternatives. The IDF is swell, i am impressed with its speed and the fact it exists, but it isn't what I need as a developer who needs to focus on other parts of the stack and has a limited amount of time. 100% Arduino compatibility is what allows rapid development with all the sensors and modules which already exist. It should be a priority for the upstream provider as opposed to making their own monolithic rival to arduino.

**Comment by WALLTECH** [comment]
@clowrey "this doesn't change the simplicity and suitability of the CurieBLE library. The hardware cost and misjudging the market for hardware needs was what's closing it down. They will leave their github up too to use." - from posts above

**Comment by kriswiner** [comment]
For product development today there is always the nRF52, which has a
passable Arduino core and BLEPeripheral support. Not ideal, but it is more
useable for products requireing BLE connectivity than the ESP32 is today. I
hope that changes. But for now, this is what I am using for my product work
this
<https://www.tindie.com/products/onehorse/nrf52832-development-board/?pt=full_prod_search>
is what I am using for my product development.

**Comment by nkolban** [comment]
I was reading the posts and something spooked me.  I try and maintain a C++ ESP-IDF library for BLE which seems to work well enough.  However, I have built my library against ESP-IDF release v2.1.  I think I saw someone post that the Arduino level of ESP-IDF is back-level.  Does anyone know of any project plans to bring ESP32 Arduino support to a level that supports the ESP-IDF 2.1?

**Comment by WALLTECH** [comment]
@kriswiner Very true, BLE peripheral is easy and can be done with a myriad of devices. BLE central however is what I'm after, and that is few and far between and what I like about CurieBLE and am hoping for from the ESP32.

**Comment by kriswiner** [comment]
Agreed. I need BLE central role devices as well. Currently, we use the
Nordic SDK to load a UART bridge solution in central role onto an nRF52. It
works great and makes for a convenient laptop dongle to pair with a nRF52
peripheral running the same Nordic UART service. But we haven't figured out
how to make it pair with more than one peripheral at a time by MAC address,
which is really what we want. BLEPeripheral currently does not support any
kind of central role.

I would expect that any easy-to-use, Arduino-compatible BLE API that does
support central role with multiple peripherals will become the standard
implementation. Could this be Curie BLE? I don't know...

**Comment by copercini** [comment]
@nkolban I got your C++ lib working on Arduino today (with all the examples of your repo), the compile time is slow on arduino IDE, so probably will need to be precompiled like others parts of core. Maybe @igrr or @me-no-dev can help with this

Now, It's just re-expose the API to match with Curie.

I don't know if is a good thing release it now as kolbanBLE or cppBLE and then when it match 100% with Curie make another release 
or
develop until it match with curie and just release the curie version

**Comment by vshymanskyy** [comment]
I have also started integrating BLE for ESP32 into [Blynk App](http://www.blynk.cc/), but then i realized that it would be much easier and better to rely on high-level API.
**CurieBLE** or **BLE Peripheral** variant would work best.
Blynk community would definitely love to create smartphone apps for this chip - we get so many requests for ESP32 support!

**Comment by nkolban** [comment]
@copercini 
That is **awesone** news.  Email me at kolban1@kolban.com or ping on IRC in channel `#ESP32` and let me know how we can collaborate (if desired).

If we (as a community) decide to go down the paths that you are working, then we have three possibilities:

1. We expose the APIs that are currently found in the cpp_utils C++ BLE library.
2. We expose the CurrieAPI building on the API currently found in the cpp_utils C++ BLE library.
3. We rework the expose API currently found in the cpp_utils C++ BLE library to natively expose the CurrieAPI.

The 4th possibility (of course) is that work is done on a complete implementation of an API (likely the CurrieAPI) starting from nothing but the raw ESP-IDF.

My personal preference would be option (1).  The thinking behind this is that it would take the least amount of effort to get it working ... and it sounds like you already have.  What we need to do next is figure out if any changes were required and then retrofit them into the common code base, likely separate it out into a "ESP32 BLE C++" project (away from the generic cpp_utils in which it lives today).  Ideally, we would want a library that can be used by **both** ESP32 programmers and Arduino programmers ... separate libraries doesn't see to satisfy any needs.

Once we have (1) ... then Arduino ESP32 programmers can start writing ESP32 BLE apps.   From there, we can test function and make sure all is working.  As experience grows, we can then achieve (2) (if needed) pretty quickly with just wrapping.  My belief is that will be performant sufficient.  However, if not performant OR the CurrieAPI doesn't map cleanly, then we can move into a gradual realm of (3).

Again ... I say all of this to try and push the ball forward.  Ideally we nominate a sub-project leader/architect.  The ideal would be the owner of Arduino on ESP32 or, failing that, a representative from Espressif.  However ... failing either of them, it could be one of us who want to see BLE on the Arduino API on ESP32.

I'll nominate myself if no-one else volunteers but if anyone else wants the job, please speak up.  We'll then arrange a quick team chat and see how far (if at all) thinking on design diverges and take it from there.  However, the ideal situation is that either the ESP32 architect @igrr or the ESP32 Arduino architect @me-no-dev takes the reigns.  From there they can either dictate what they want to see happen or delegate down to one of us in this issue/community.

**Comment by me-no-dev** [comment]
I personally prefer to only take other API as a general guide and have something implemented that makes more sense with the lower IDF API. All layers of classes and abstractions need to be properly written so it is easy to implement all higher level functions and not break one another. My first thought is what will users do and break. Not so important is which API to adhere to, because much more users will jump on the ESP32 than any other such platform.

**Comment by copercini** [comment]
Let’s show some code: in attachment is the functional cpp_utils C++ port as Arduino library (it’s still missing some examples, but the intent is more discussion of APIs)
[BLE.zip](https://github.com/espressif/arduino-esp32/files/1203043/BLE.zip)


My thoughts about it:

#### Compatibility between IDF and Arduino:
- Log system is quite different but there must be some way to unify both 

#### Compatibility between c++BLE and Arduino:
- Don’t expose all code of callback classes in the sketch (maybe use something like [SetEventHandler](https://www.arduino.cc/en/Reference/BLEPeripheralSetEventHandler) )
- Simplify the way how objects are instantiated, like:
`BLEService serviceName("19B10000-E8F2-537E-4F6C-D104768A1214");`
Instead of 
`BLEService *serviceName = pServer->createService(BLEUUID(("19B10000-E8F2-537E-4F6C-D104768A1214"));`
- ESP32 have some extra options like active/passive scan



#### Issues:
- BLE is so huge: a simple example takes 80% of flash partition for programs
- Compilation is slow on Arduino IDE (takes about 5 minutes here)

**Comment by copercini** [comment]
@me-no-dev 

> I personally prefer to only take other API as a general guide 

It's CurieBLE

> and have something implemented that makes more sense with the lower IDF API. 

It's cpp_utils from @nkolban 

> Not so important is which API to adhere to, because much more users will jump on the ESP32 than any other such platform.

cpp_utils and CurieBLE are almost close (some code are in the .zip above), personally I vote to change a bit of cpp_utils APIs to match with CurieBLE (which already has documentation and examples ready) and keep the compatibility between IDF and Arduino to use the same library in both

**Comment by d3cline** [comment]
TYVM @copercini I am testing the library now. I had to do a git pull to the latest hardware/arduino-esp32 if anyone else runs into compile errors and this did not occur to them.

**Comment by d3cline** [comment]
I have tested the 4 examples and BLE_scan seems to function without error. The others compile but I get a crash in the serial monitor different for each. I could post the errors and work through them on a better thread when/if one is created.

**Comment by d3cline** [comment]
As for feedback. I am impressed by the simplicity already, great work. Much better than 500+ lines, and it does compile on Arduino 1.8.0 on Ubuntu. Going to try to get it to read some I2C and send it. This all looks very promising.

**Comment by nkolban** [comment]
I had some good chats with Mr @copercini and understand what he did to enable BLE within the Arduino environment.  I was able to recreate and it looks great.  At the highest level he has determined the recipe for building the `cpp_utils` BLE library for the ESP32 Arduino environment.

The next stage was to build a "Makefile" that will be part of the `cpp_utils` that can build an Arduino BLE.zip library file from the `ccp_utils` source that is ready for distribution as an Arduino BLE library.  This has been done and is now being better tested.  When this is complete, it will allow maintainers of the eventual Arduino BLE library to repeatedly build the ZIP without manual or mistaken tasks.

One of the design goals is to ensure that the BLE library is as close to the same code for ESP-IDF as for Arduino and so far we have zero discrepancies between the two allowing us to have an identical code base without any conditional compilation logic.  However, the one snag (and its a small one) is that the BLE library on ESP-IDF allows one to debug the internals using the ESP32 ESP_LOGx() macros.  These macros are disabled in the Arduino environment.  They *could* be quite easily enabled.  Issue #565 asks for permission to change the code base of the Arduino environment ( a few lines) to enable ESP_LOGx() macros in the Arduino environment.  At that point we could have what would appear to be 100% code affinity between BLE for ESP-IDF and BLE for Arduino.

**Comment by nkolban** [comment]
Work continues on the BLE library polish for Arduino.  A simple BLE scan worked without incident.  No issue.  Next I tried to create a BLE Server but that's where things started to get sticky.  The current problem is logged as issue #568 which results in a rebooting exception.   My guess work is loosely the following:

In ESP-IDF programming, we have a utility called "make menuconfig" which allows us to change the characteristics of the environment in which an ESP-IDF application will run.  When we run "make menuconfig", the result is a configuration file that eventually becomes "sdkconfig.h".  If we look at the ESP32 Arduino environment, we don't have "make menuconfig" instead we have a pre-supplied "sdkconfig.h".  See:

https://github.com/espressif/arduino-esp32/tree/master/tools/sdk/include/config

and this is where things slow down and become painstaking.  What I believe I need to do is compare, line by line, the settings "present" in the pre-supplied "sdkconfig.h" with the settings I understand I want/need for BLE to work in ESP-IDF using "make menuconfig".  Without the menu driven "make menuconfig" it isn't "easy" to see what is set, what is not set, what are defaults and what are customized.  My gut is saying that there is something in the Arduino environment that is distinct from the the default/customized ESP-IDF environment.   As a loose example, I have found that when working in ESP-IDF, the default BLE task stack size is not large enough and I usually have to bump it up from the default 2K to about 8K.  The default in the Arduino sdkconfig remains at 2K (and yes, I tried that but that hasn't resolved the core issue).  Unfortunately, it looks like I have to grunge through the lines one by one looking for the distinctions ... this may be a good opportunity for a tool or script.

**Comment by MrSurly** [comment]
@nkolban 

I went through the same thing w/ BLE for MicroPython.  You can see the changes necessary in https://github.com/MrSurly/micropython-esp32/tree/dev-bluetooth, by comparing `esp32/sdkconfig.h` to the version found in the `esp32` branch.  The BT differences are summed up in a handy `#if` about halfway down.

I seem to remember two big ones: `CONFIG_BT_RESERVE_DRAM`, and the fact that the available heap went from about 100K to about 48K.  Not sure the latter will affect Arduino much.  In MicroPython, most of the available heap is taken for use as the Python heap.

```diff
diff --git a/esp32/sdkconfig.h b/esp32/sdkconfig.h
index de8630e..81587e0 100644
--- a/esp32/sdkconfig.h
+++ b/esp32/sdkconfig.h
@@ -1,9 +1,9 @@
 #define CONFIG_TRACEMEM_RESERVE_DRAM 0x0
-#define CONFIG_BT_RESERVE_DRAM 0x0
 #define CONFIG_ULP_COPROC_RESERVE_MEM 0
 #define CONFIG_PHY_DATA_OFFSET 0xf000
 #define CONFIG_APP_OFFSET 0x10000
 
+
 #define CONFIG_SPI_FLASH_ROM_DRIVER_PATCH 1
 #define CONFIG_FOUR_UNIVERSAL_MAC_ADDRESS 1
 #define CONFIG_NUMBER_OF_UNIVERSAL_MAC_ADDRESS 4
@@ -51,7 +51,6 @@
 #define CONFIG_FREERTOS_ASSERT_ON_UNTESTED_FUNCTION 1
 #define CONFIG_FREERTOS_CHECK_STACKOVERFLOW_NONE 1
 #define CONFIG_FREERTOS_THREAD_LOCAL_STORAGE_POINTERS 2
-#define CONFIG_FREERTOS_IDLE_TASK_STACKSIZE 1024
 #define CONFIG_FREERTOS_ISR_STACKSIZE 1536
 #define CONFIG_FREERTOS_BREAK_ON_SCHEDULER_START_JTAG 1
 #define CONFIG_FREERTOS_MAX_TASK_NAME_LEN 16
@@ -59,7 +58,6 @@
 #define CONFIG_ENABLE_STATIC_TASK_CLEAN_UP_HOOK 1
 
 #define CONFIG_MAIN_TASK_STACK_SIZE 4096
-#define CONFIG_IPC_TASK_STACK_SIZE 1024
 #define CONFIG_BTC_TASK_STACK_SIZE 3072
 #define CONFIG_SYSTEM_EVENT_TASK_STACK_SIZE 4096
 #define CONFIG_SYSTEM_EVENT_QUEUE_SIZE 32
@@ -70,9 +68,24 @@
 #define CONFIG_NEWLIB_STDOUT_ADDCR 1
 #define CONFIG_PHY_ENABLED 1
 #define CONFIG_WIFI_ENABLED 1
+#define CONFIG_SMP_ENABLE 1
 #define CONFIG_OPTIMIZATION_LEVEL_DEBUG 1
 #define CONFIG_MEMMAP_SMP 1
 
+#if (MICROPY_ESP32_BLUETOOTH)
+#   define CONFIG_BT_ENABLED 1
+#   define CONFIG_BLUEDROID_ENABLED 1
+#   define CONFIG_GATTS_ENABLE 1
+#   define CONFIG_BT_DRAM_RELEASE 1
+#   define CONFIG_GATTC_ENABLE 1
+#   define CONFIG_BT_RESERVE_DRAM 0x10000
+#   define CONFIG_BTDM_CONTROLLER_RUN_CPU 0
+#   define CONFIG_BT_ACL_CONNECTIONS 4
+#   define CONFIG_BLE_SMP_ENABLE 1
+#else
+#   define CONFIG_BT_RESERVE_DRAM 0
+#endif
+
 #define CONFIG_PARTITION_TABLE_SINGLE_APP 1
 #define CONFIG_PARTITION_TABLE_FILENAME "partitions_singleapp.csv"
 #define CONFIG_PARTITION_TABLE_CUSTOM_APP_BIN_OFFSET 0x10000
@@ -110,3 +123,6 @@
 
 #define CONFIG_TOOLPREFIX "xtensa-esp32-elf-"
 #define CONFIG_PYTHON "python2"
+
+
+
```

**Comment by nkolban** [comment]
Update:
First a question ... given the chatter on this issue, do we want to create a forum thread for our discussions on BLE support for ESP32 Arduino or do we wish to continue using this Git issue?   My vote would be a thread but we'll let majority decide.   Here is an online poll we can use:

http://www.easypolls.net/poll.html?p=598a9015e4b05211e16ff0e6

Now for the update.  At the end of yesterday we found that we were running into problems building and running BLE servers.  After experimentation, it was found that this was caused as a result of an experiment that switched on ESP_LOGx debugging.   This is the diagnostics technology that we leave on during development to visualize what is going on inside the BLE libraries.  Normally, users wouldn't need to run with that on.  It turns out that our hope that we had a recipe to enable ESP_LOGx within Arduino weren't as fruitful as we'd believed.  When we switched off the ESP_LOGx debugging (which is the default), the crashes went away and all appears to function.  So ... hooray.

Mr @copercini made some fantastic changes to the project to make specification of service and characteristic UUIDs much easier.  These are now merged with master.

As changes will likely be coming thick and fast for a while and we are still months away from anything we might term a "release", we wanted to make sure that anyone who wished, could get a "build" from the latest code anytime they wanted.  As such, we now have a written recipe for building such that should hopefully be pretty straight forward.  It has been written up here:

https://github.com/nkolban/esp32-snippets/blob/master/cpp_utils/ArduinoBLE.md

If you decide to play with the builds of the BLE library, it will be assumed that you are using the latest build created from these instructions.  This will give us "one source of truth" on what code is in effect at any given time.

**Comment by me-no-dev** [comment]
I prefer you stick to using this github issue, or I will not be able to follow the development. As you @nkolban know, I am not big on forums, nor I have the time to go through them.

**Comment by me-no-dev** [comment]
```#define CONFIG_BT_RESERVE_DRAM 0x10000``` seems a bit overkill ;) let's not waste half our memory on BT ;) not everyone uses it.

**Comment by nkolban** [comment]
@me-no-dev When writing a BLE application using ESP-IDF, we commonly make changes to the sdkconfig using `make menuconfig`.  Common things that we do include increasing the stack size for the BLE task.  If we assume that the default settings for the Arduino environment aren't sufficient for some projects that would leverage BLE on the Arduino on ESP32, do you have a suggestion on design or practices that we should use?

**Comment by zencuke** [comment]
Sorry if this is off point but how do you track maximum stack usage. I know how I've done it in other platforms but is there something built in I could just turn on?

**Comment by nkolban** [comment]
As I understand it, each FreeRTOS task has its own stack that is of fixed size and specified when the task is created.  FreeRTOS does have "high water mark" measurement to record (and report) the highest value of stack space used ... however I haven't looked at that area in a long while.  The clue (to me) that we had to increase the stack space were simple "crashes" when it was the default (too low).  It appears that the FreeRTOS stack management has mechanisms to detect when a request to push to the stack would exceed the allocated amount (there may be a couple of techniques including markers (a canary) that get checked to see if they have been overwritten).  I was able to get through normal usage a repeatable story that would cause a crash when writing "some" BLE applications in ESP-IDF that indicate that we we had run out of BLE task stack space and when I bumped it up (using an arbitrary value of 8000 as opposed to 2048), I no longer had the issue.  In an ideal world, we should be able to predict or create a formula that will tell us the stack size we need to allocate for given set of use cases so that we can allocate the stack size without wasting storage.  However, I don't know what that is yet and determining a technique as you suggest of being able to actually measure stack space usage sounds like something very useful.

**Comment by MrSurly** [comment]
>#define CONFIG_BT_RESERVE_DRAM 0x10000 seems a bit overkill ;) let's not waste half our memory on BT ;) not everyone uses it.

@me-no-dev: It's all-or-nothing, per @Spritetm of Espressif: https://github.com/espressif/esp-idf/issues/487#issuecomment-291731836

For MicroPython, the compilation of BLE will be configurable.

**Comment by lonerzzz** [comment]
I have been creating FreeRTOS task threads with their own stack allocations and this does make sense as a way to go in terms of allocating the right amount of stack for BLE while not affecting the Arduino default stack size. One question that arises from this is whether anything about the BLE implementations would prevent them running in a separate task thread. I have encountered a few issue running some library code on threads different than the main Arduino task thread.

**Comment by copercini** [comment]
My unique issue until now is some out of stack in Btc_task https://github.com/espressif/arduino-esp32/issues/567

So I think increasing it a bit, we can run the things without a big reservation of heap

**Comment by nkolban** [comment]
@lonerzzz 

I probably haven't been at all clear.  Within the implementation of BLE within ESP-IDF, there is a task that is created behind the scenes.  This is task that is owned and managed by ESP-IDF.  The only exposure we have to it is as shown here:

![image](https://user-images.githubusercontent.com/3421732/29142933-13f38ecc-7d19-11e7-8ec4-87710eb22b5b.png)

here we see the "make menuconfig" settings for Bluetooth in ESP-IDF.  See that there is a stack size setting for a the task.  This can't be set at "runtime" but rather needs to be set during compilation.  In an ESP-IDF environment, we would run "make menuconfig" ... however that doesn't appear to be an option in Arduino land.  We can potentially edit a file called "sdkconfig.h" buried deep within the Arduino implementation ... but there might be an opportunity to pre-define values for BLE ... which would then waste resources if the Arduino app wasn't using BLE.

**Comment by zencuke** [comment]
@nkolban Thanks. Both ways of detecting stack problems are very respectable. I'll check the FreeRTOS doc for usage. Dynamic tracking is expensive but having a "max used" that can be checked is good. It is helpful to set the stacks high for development then determine actual need before release. The trick is to remember.  ;-)

**Comment by lonerzzz** [comment]
@nkolban Thanks for the clarification. Didn't realize the task thread ownership

**Comment by srinadhpotnuru** [comment]
Has anyone used the library developed by @copercini to develop a BLE central  and peripheral on ESP-32

**Comment by nkolban** [comment]
We'll see if we can't publish some sample as Arduino .ino files ... in the meantime here are samples in ESP-IDF that should be VERY close to what will be the end result as those .ino files.

https://github.com/nkolban/esp32-snippets/tree/master/cpp_utils/tests/BLE%20Tests

**Comment by mxwang101** [comment]
Thanks，waiting for……
On 08/16/2017 21:25, Neil Kolban wrote:

We'll see if we can't publish some sample as Arduino .ino files ... in the meantime here are samples in ESP-IDF that should be VERY close to what will be the end result as those .ino files.

https://github.com/nkolban/esp32-snippets/tree/master/cpp_utils/tests/BLE%20Tests

**Comment by mxwang101** [comment]
We do need some example code of BLE central which can read and write the information to the servers.
On 08/16/2017 21:25, Neil Kolban wrote:

We'll see if we can't publish some sample as Arduino .ino files ... in the meantime here are samples in ESP-IDF that should be VERY close to what will be the end result as those .ino files.

https://github.com/nkolban/esp32-snippets/tree/master/cpp_utils/tests/BLE%20Tests

**Comment by copercini** [comment]
About the four first examples:
The write, server and notify are working without know issues
Only the scan depends of https://github.com/espressif/arduino-esp32/pull/583

 I think they can be merged to kolban repository

---
About client examples still need a bit more of work....

**Comment by telamon** [comment]
Not sure if i should share this info here but I've tested the current BLE code and can confirm that server with write and read + callbacks works.
But on the other hand I noticed that all calls to `analogRead()` breaks and start outputting a fixed value of  `4095` (adc max value).
I've managed to trace this behaviour down to the function `esp_bt_controller_enable` from the esp-idf api: http://esp-idf.readthedocs.io/en/latest/api-reference/bluetooth/controller_vhci.html?highlight=esp_bt_controller_enable#_CPPv224esp_bt_controller_enable13esp_bt_mode_t
As soon as that function gets called all analog readings get boring :(

Does anyone know where this issue belongs?

**Comment by neilbags** [comment]
One of the ADCs is unusable when Bluetooth or WiFi is active. Try the other set of pins.

**Comment by telamon** [comment]
That explains it, do you know if it's ADC1 or ADC2 that gets unusable?

**Comment by nkolban** [comment]
Caution: ESP-IDF master appears to break BLE.  I've been finding that when we use the ESP-IDF master image, BLE breaks.  See this issue https://github.com/espressif/esp-idf/issues/857

The version of ESP-IDF used in Arduino land is one that is packaged with the Arduino development environment so at the moment, we are using what ever has been selected by the Arduino environment developers ... however we need to make sure that if we want to continue using BLE on the Arduino platform that we ask them to ensure that they don't package with ESP-IDF master until the references issue is resolved.

**Comment by me-no-dev** [comment]
ok, so no IDF update today?

**Comment by nkolban** [comment]
Howdy @me-no-dev 
Can you share what your plans/schedules are with respect to refreshes of the project to use later ESP-IDF builds?  Obviously the decision will be yours to determine when is the appropriate time to use different builds.  I don't know if you constrain the project today to use only release builds (2.1 is the latest I believe) or whether the project uses "best so far" (i.e. ESP-IDF master).   Can you let us know the process you use to determine with ESP-IDF you use?

**Comment by me-no-dev** [comment]
I try to update every other week or more often if there is a major change in IDF that could prevent Arduino from working as a component. I saw BT changed and some other stuff since the last build, so I thought of running an update (there are a couple of other things to come next time also), but since you say that BT is now broken,  will hold on that :)

**Comment by nkolban** [comment]
Sample/Example BLE sketches for Arduino - Mr @copercini has written and tested some great examples for using the BLE classes in Arduino and committed them to the project.  They can be found here:

https://github.com/nkolban/esp32-snippets/tree/master/cpp_utils/tests/BLETests/Arduino

He has also augmented the Makefile for building the ESP32_BLE.zip Arduino library such that we now have example integrated into the Arduino IDE under the "Examples" menu.  Awesome stuff.

A bug that we had been chasing for some time was also alleviated.  The bug manifests itself as an exception that references "Canary storage" and was non-deterministic (sometimes it would happen and sometimes it wouldn't).  This has been resolved to what would appear to be a race condition.  We had been initializing BLE subsystems with `BLEDriver::init()` and then calling BLE functions immediately.  We found that if we place a 200 msec paused following `init()`, the exception wen't away.  We have internally added that to the end of the `init` function such that it will now take 200 msecs more to complete ... but the good news is that you don't have to do anything special.

**Comment by me-no-dev** [comment]
there should be a reason behind the need of that 200ms. Maybe it takes a bit to spin up a BT task. How did you find that value and what else have you tried? 200ms sounds like ages :D

**Comment by nkolban** [comment]
The 200ms was a guess.  The way we found it was that my colleague was testing and for him it was failing 95% of the time.  For me, it was failing, 20% of the time.  He was using the sdkconfig that comes with Arduino while I had mine customized.  We compared the two sdkconfig.h files and by a process of trial and error found the one change that resulted in mine working "more often" than his.  The distinction was that I had default logging turned up to 5 as opposed to the default of 1.  Thinking through what change "logging" might introduce it dawned on me that I was increasing path length and possibly slowing things down.  On a hunch a tried a delay of a couple of seconds after BLE initialization before the first BLE request and that worked 100% of the time.  I then pulled a small number out of the air (200 msecs) and that too worked 100% of the time and stopped there.  In theory, we could keep trying smaller and smaller numbers but 200 msecs penalty is "ok".  The BLE initialization happens one time.

The source statement where we are now delaying is here:

https://github.com/nkolban/esp32-snippets/blob/master/cpp_utils/BLEDevice.cpp#L204

**Comment by me-no-dev** [comment]
ok, let's try to figure out what could be causing the need for this delay. @igrr can we have the BT guys take a look? Maybe we need some sort of "event" that BT is ready?

**Comment by nkolban** [comment]
If there is anything I can do to assist, I'll obviously be delighted.  The puzzle was independently verified on both my machine and that of mr @copercini so that bodes well for a recreate within Espressif.  The recreate can be had by building an Arduino environment, adding in the BLE library and running the BLEScan sample (that's what we used for testing).  Remove or uncomment the delay for 200msecs that we have added as a circumvention and then it should manifest.  If needed, I'll be happy to demonstrate via screen share.

**Comment by copercini** [comment]
The most critical is the BLE_scan example, the minimal delay for work fine is ~100ms
BLE_scan example needs ~10ms
and the others run without delay without problems here

**Comment by MarosValter** [comment]
Hello guys, I've been watching this thread for a while and finally wanted to try out BLE. I've build cpp_utils library (manully, there were some problems with make on windows) and updated esp arduino core. But I was unable to make work write and notify examples (scan works fine).
It looks like the problem you mentioned up with rebooting, but the issue was not the bluetooth stack size (which was incremented in a recent commit) nor the ESP_LOGx macros (I tried to comment them, just to be sure).
I use ESP32 Thing (https://www.sparkfun.com/products/13907). After I upload the program, board keeps rebooting forever and it looks like it happens after BLEService *pService = pServer->createService(SERVICE_UUID);

Here is ouput from serial monitor:
ets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0010,len:4
load:0x3fff0014,len:588
load:0x40078000,len:0
load:0x40078000,len:9768
entry 0x400789b8
abort() was called at PC 0x400d880b on core 0

Backtrace: 0x400887fc:0x3ffde010 0x400888fb:0x3ffde030 0x400d880b:0x3ffde050 0x400d9e7a:0x3ffde070 0x400da006:0x3ffde090 0x400da351:0x3ffde0d0 0x400da391:0x3ffde100 0x400da3a1:0x3ffde120 0x400d6252:0x3ffde140 0x400d4ecd:0x3ffde160 0x400d3e45:0x3ffde1a0 0x400f2cfd:0x3ffde1c0 0x400eef04:0x3ffde200

Rebooting...

Can you help me please resolve this issue ? I have no clue, where to start...

Anyway thank you for your effort, you made an excellent work!
Thanks

**Comment by copercini** [comment]
@MarosValter Start decoding the backtrace with https://github.com/me-no-dev/EspExceptionDecoder

**Comment by MarosValter** [comment]
@copercini  Ok, this is what I got:

0x400887fc: invoke_abort at /Users/ficeto/Desktop/ESP32/ESP32/esp-idf-public/components/esp32/./panic.c line 546
0x400888fb: abort at /Users/ficeto/Desktop/ESP32/ESP32/esp-idf-public/components/esp32/./panic.c line 546
0x400d880b: std::__throw_length_error(char const*) at ?? line ?
0x400d9e7a: std::__cxx11::basic_string  , std::allocator  >::_M_create(unsigned int&, unsigned int) at /home/ivan/e/crosstool-NG/.build/xtensa-esp32-elf/build/build-cc-gcc-final/xtensa-esp32-elf/libstdc++-v3/include/bits/basic_string.h line 2200
0x400da006: std::__cxx11::basic_string  , std::allocator  >::_M_mutate(unsigned int, unsigned int, char const*, unsigned int) at /home/ivan/e/crosstool-NG/.build/xtensa-esp32-elf/build/build-cc-gcc-final/xtensa-esp32-elf/libstdc++-v3/include/bits/basic_string.h line 2200
0x400da351: std::__cxx11::basic_string  , std::allocator  >::_M_replace(unsigned int, unsigned int, char const*, unsigned int) at /home/ivan/e/crosstool-NG/.build/xtensa-esp32-elf/build/build-cc-gcc-final/xtensa-esp32-elf/libstdc++-v3/include/bits/basic_string.h line 2200
0x400da391: std::__cxx11::basic_string  , std::allocator  >::assign(char const*) at /home/ivan/e/crosstool-NG/.build/xtensa-esp32-elf/build/build-cc-gcc-final/xtensa-esp32-elf/libstdc++-v3/include/bits/basic_string.h line 2200
0x400da3a1: std::__cxx11::basic_string  , std::allocator  >::operator=(char const*) at /home/ivan/e/crosstool-NG/.build/xtensa-esp32-elf/build/build-cc-gcc-final/xtensa-esp32-elf/libstdc++-v3/include/bits/basic_string.h line 2200
0x400d6252: FreeRTOS::Semaphore::give() at C:\Projects\Arduino\libraries\ESP32_BLE_master\src/FreeRTOS.cpp line 126
0x400d4ecd: BLEServer::handleGATTServerEvent(esp_gatts_cb_event_t, unsigned char, esp_ble_gatts_cb_param_t*) at C:\Projects\Arduino\libraries\ESP32_BLE_master\src/BLEServer.cpp line 47
0x400d3e45: BLEDevice::gattServerEventHandler(esp_gatts_cb_event_t, unsigned char, esp_ble_gatts_cb_param_t*) at C:\Projects\Arduino\libraries\ESP32_BLE_master\src/BLEDevice.cpp line 43
0x400f2cfd: btc_gatts_cb_to_app at /Users/ficeto/Desktop/ESP32/ESP32/esp-idf-public/components/bt/bluedroid/btc/profile/std/gatt/btc_gatts.c line 53
:  (inlined by) btc_gatts_cb_handler at /Users/ficeto/Desktop/ESP32/ESP32/esp-idf-public/components/bt/bluedroid/btc/profile/std/gatt/btc_gatts.c line 730
0x400eef04: btc_task at /Users/ficeto/Desktop/ESP32/ESP32/esp-idf-public/components/bt/bluedroid/btc/core/btc_task.c line 85

**Comment by nkolban** [comment]
Howdy  @MarosValter 
If you are willing, I'd like to try a copy of the sketch in my environment.  Would you be able to post a copy via pastebin and attach a link?

**Comment by MarosValter** [comment]
Hi @nkolban,
actually, I was kinda wrong, write example is working very well, problems are with server and notify examples. There are exactly the ones provided in your cpp_utils, but here is copy, if it will be helpful:
server: https://pastebin.com/SAYjDPiq
notify: https://pastebin.com/tAQk1TaN

**Comment by clowrey** [comment]
Both Server and Notify are working great for me - all on newest Adruino core and libs from nkolban. 

I have combined the notify and write examples into a sort of UART example, it actually uses the standard UUID for a UART and works with common Android apps that talk to bluetooth UARTs. The actual input/output "print" functions are not written yet but all the data is there. I have just made a pull request here: https://github.com/nkolban/esp32-snippets/pull/43

**Comment by nkolban** [comment]
@MarosValter 
When we have non-trivially reproducible puzzles, those (unfortunately) take longer to figure out.  Here are some of the more common questions we ask:

1. Are you following these instructions? - https://github.com/nkolban/esp32-snippets/blob/master/cpp_utils/ArduinoBLE.md
2. Are you using the latest version of the ESP32-Arduino libraries?
3. Can you paste.bin your local version of sdkconfig.h?

If I were to make a wild guess based on the symptoms reported, we see that the failure is occurring within `std::__cxx11::basic_string , std::allocator >::_M_create(unsigned int&, unsigned int)` and that *feels* like a memory allocation error.   Since the code is running on the stack for the Blue Tooth stack which is governed by CONFIG_BTC_TASK_STACK_SIZE, I have a feel that the amount of stack space may be too low.

@clowrey 
MANY thanks for the new example. Merged and committed.

**Comment by MarosValter** [comment]
@nkolban
1. yes, as I wrote in my first post, I followed those instructions but I was unable to make the library using makefile, so I manually created directory structure and then manually copied source files, library properties as well as examples
2. yes. I updated esp arduino libraries from their repo
3. https://pastebin.com/fMZ0akXL

Thank you for your help

**Comment by chenkaiyao** [comment]
@igrr @me-no-dev 
can youtell me the deadline of the function of ble?

**Comment by nkolban** [comment]
@chenkaiyao 
My belief is that it is up to us as the community to put together the BLE wrappers for Arduino.  Let us recap.

First, BLE support appears to be 100% present in ESP-IDF.  ESP-IDF is the foundation platform on which all ESP32 applications are built ... including much of the ESP32 Arduino libraries.  What this means is that any implementation BLE functions for Arduino will be super likely to be based on top of the BLE APIs found in ESP-IDF.  The original post by the Espressif architect @igrr  and the ESP32 Arduino owner @me-no-dev was to solicit our opinion on which API we wanted to use to expose BLE functions in Arduino.  As I understand it, that was as much a "requirements gathering" exercise as anything else.

Unrelated to the Arduino story, another sub-project was attempting to encapsulate the BLE APIs found in ESP-IDF into a set of C++ classes.  Since the "model" of Arduino is linked closely to C++ programming, some folks said "I wonder if the C++ BLE sub-project could run within the Arduino framework" ... and low-and-behold, it could.  This was then packaged into more of an Arduino friendly context (a ZIP file that could be imported into Arduino IDE).  And that is where we are today.  We have *an* implementation of BLE functions that will work within the Arduino library and IDE for ESP32.  Will it be the **final** one?  Possibly not ... but it would seem a shame for others to start again 100% from scratch.  The C++ sub-project is not locked in stone.  If we, as a community, want changes made, APIs added or interfaces modified ... then that might be an easier place to start that starting from scratch.

**Comment by mxwang101** [comment]
I think if have something like "CurieBLE" , will be good.  

I do used CurieBLE with Arduino/Genuino 101. They are perfect.

**Comment by WALLTECH** [comment]
@nkolban How would one use the class functions in an ino sketch? Can the classes be curieBLE formatted for ease of use, and could this not then be the BLE library complete (for now)?

**Comment by nkolban** [comment]
There are some docs on the ESP-IDF C++ BLE classes available here:

https://github.com/nkolban/esp32-snippets/blob/master/Documentation/BLE%20C%2B%2B%20Guide.pdf

and some instructions for building the ZIP file from the source that can then be imported into the Arduino IDE available here:

https://github.com/nkolban/esp32-snippets/blob/master/cpp_utils/ArduinoBLE.md

Some sample Arduino .ino sketches illustrating use are available here:

https://github.com/nkolban/esp32-snippets/tree/master/cpp_utils/tests/BLETests/Arduino

As for the "curieBLE" story ... the ESP-IDF BLE C++ classes were "made up" as they were being designed and developed.  Too little attention was spent on prior art interfaces and the "curieBLE" interface description wasn't considered as a template against which to design.  The good news is that it appears that (accidentally) the same thought process in "curieBLE" and "BLE C++ in cpp_utils" were close to each other.

So now we have the following options:

1. We can live with what we have and use the existing API set that has been coded today.
2. We can re-code the API in "BLE C++ in cpp_utils" to implement the "curieBLE" interfaces.
3. We can "wrap" the "BLE C++ in cpp_utils" to expose the "curieBLE".
4. We can start again from scratch and implement "curieBLE" using ESP-IDF from the ground up.

See also the discussion that was had about a month ago on this very topic.

**Comment by WALLTECH** [comment]
Whatever is easiest I'd say. Then we have something working we can optimize or recode for efficiency. Would that be wrapping to expose curieBLE?

**Comment by nkolban** [comment]
Howdy John,

From my perspective, the easiest would be option (1) ... which is to ignore the "curieBLE" API and use the API documented here ...

https://github.com/nkolban/esp32-snippets/blob/master/Documentation/BLE%20C%2B%2B%20Guide.pdf

It seems that this API works in Arduino today.  If we don't want to use that API then the next question is "Which API?" ... the answer to that seems to have been answered in this issue already with the votes for "curieBLE".  I for one am willing to pitch in and provide a "curieBLE" API for the "BLE C++ in cpp_utils" ... and here's the **but** ... only if there is true and verifiable demand for such.  That sounds like a strange thought ... but here's where its coming from.  In my travels in Open Source land, I find folks really keen to have a piece of function and in the end when it is delivered, end up not using it.  I'm all for pushing the boundaries of community based software ... but only if there is demonstrable demand and others who are willing to pitch in ... even if it is only testing or requirements gathering or documentation.   I'm hearing mixed signals on "curieBLE" API vs "BLE support for ESP32 Arduino".  By almost accident, we have satisfied something that could be loosely considered "BLE support for ESP32 Arduino" but yet when we look at download logs and questions and comments ... the feedback is almost nill.  Does that mean it works flawlessly straight away for everyone?  Does that mean that everyone is waiting for the "curieBLE API"?  Does it mean that interest in BLE support for ESP32 Arduino isn't really that high?

**Comment by WALLTECH** [comment]
@nkolban I completely understand, and agree. It seems like a lot of people have posted here about curieBLE, and especially as the curie is now discontinued, or will be in a couple of weeks, a drop in replacement would be quite useful in the ESP32. I for one plan on using this to combine the ESP32 and curie combination I have going for IoT in prototypes, and make it a multi thousand unit product. I could of course spend some time adapting it for this API, but a no modifications switch would almost be too good to be true for that many lines of code :) I haven't much experience with behind the scenes of Arduino, but would like to help if I can. All in all, I'll be using it for a long time, and very widely if that helps.

**Comment by d3cline** [comment]
+1 Working solution whatever API. The simplified approach is enough for me.

**Comment by JeroenBeemster** [comment]
A lot of thoughts over here. May be it is time move on? 
From an average Arduino user point of view we have currently 'SimpleBLEdevice' in the examples.  (I know it is not fully true). This is what I used until now. I have a solution that runs and break every 24 hours. I am waiting for something easier to maintain.  imho: nkolban did a great job. But for a widely use in the Arduino world, the 'average' user will still find it a little bit unreadable. Don't get me wrong nkolban you did a very good job. 
So my vote: curieBLE. But any decision today is better than tomorrow.

**Comment by jcmojj** [comment]
@nkolban <https://github.com/nkolban> I understand that you see a lot of
people that asked for "curieBLE" and some people just want the "BLE support
for ESP32". I finally heard, after reading all those posts, that you
finally have something close to "BLE support for ESP32", but it looks like
few people use it already. You don't know if it happened because few people
is interested in BLE or because you don't have curieBLE. From my point of
view and actually the reason why I didn't use it yet is that I am waiting
for something that will not modify. I need as soon as possible to start
using it, but I can't take the risk of changing and loose a lot of stuff.
More than that, I have a question that wasn't discussed here yet. I read
some posts ago that you used a lot of memory for tests because you don't
know for sure how much memory will it use. You tried few memory and it
crashed. Do you think that if you reduce the memory usage for an acceptable
size that allows BLE to work on ESP32, we would have enough space yet to
use OTA and WiFi together with our own codes in the esp32 dev kit. I image
that the biggest part of users are using it because it is easier and those
are the basic apis that most people will probably use together. If the
answer is yes, I will have enough space to use it all and our code is ready
to use, and also say that the code we start using won't change (at least
the interface of if), everybody can start using it without the fear of
losing the code or at the end don't have space to do basic stuff. I am the
average Arduino user and this is my point of view. If the code is a bit
unreadable, all we need is good documentation. I prefer a code that works
and we will be able to use in two week than one simpler that will take 2
months to be ready.

**Comment by neilbags** [comment]
I made an example sketch for BLE-MIDI, based on @copercini's BLE_notify example. Currently works with Linux clients (Bluez >=5.44) and Windows 10 Creators Update 1.  Haven't got a mac so I can't test it with OS X. 

https://github.com/neilbags/arduino-esp32-BLE-MIDI

Feel free to use this or adapt it. Thanks for all your hard work!

**Comment by ayavilevich** [comment]
@nkolban Hi, some feedback.
I have plans to use BLE on Arduino-ESP32, so I am monitoring the progress but not using the library right now. Hence no feedback. I assume there are others who are in the same situation. It might take time for the library to gain users. It is just a natural thing.
Personally I don't mind if this is the current API or the CurieBLE API. What matters is that I can use it to implement my BLE logic (client/server, etc).
Thanks everybody for the work done on this library so far. Looking forward to contribute in the future.

**Comment by nkolban** [comment]
@neilbags, @ayavilevich 

Good to hear that there may be some value in the code.  The Github repository for the BLE library currently lives here:

https://github.com/nkolban/esp32-snippets

So if we have any technical problems with the library or request for enhancement or correction, please feel very free to post issues there for tracking and resolution.  My thinking is that this Git hub issue (the one we are reading now) is as much a meta discussion on which BLE library we will use for the ESP32 Arduino environment.  If you are working with the "BLE C++ library in cpp_utils" and find a problem (eg. divided by zero when message size is 512), I'm not sure that the majority of readers of this thread/issue will be interested.  So for normal and expected technical issues, I suggest raising those as issues on the Github repo listed above.

**Comment by DonWasabi** [comment]
@nkolban Hi, and another feedback: I am visiting this issue thread nearly daily and from what i read the API seems to be kinda stable, so now seems to be a good time to try out the Arduino BLE functionality :) Thanks for your efforts.

**Comment by WALLTECH** [comment]
@nkolban So is central scanning, connecting, and writing/reading from peripherals working in your snippets? There isn't an example for it, but that's what I need from BLE.

**Comment by nkolban** [comment]
@WALLTECH Yes sir.  Those functions should be present.  Have a look at these samples here:

https://github.com/nkolban/esp32-snippets/tree/master/cpp_utils/tests/BLETests/Arduino

If there are additional samples you would like to see that aren't present, create an issue here and we'll get to work building them out.

https://github.com/nkolban/esp32-snippets/issues

**Comment by WALLTECH** [comment]
Ok, because all the ones in there seem to be peripheral and not central, would I just modify the cpp example for it to the arduino format like the others? A BLE Central scan for device, connect, enable notifications or other characteristic and read/write the value would be really useful.

**Comment by WALLTECH** [comment]
Attempted it and can't decide what's required and not and the client example is a bit too different for me from the others to translate it. I posted in the snippets repo about a client/central example.

**Comment by DonWasabi** [comment]
Hi, are there some news of the status of the (official?) curie BLE arduino library?

**Comment by haosheng-huang** [comment]
Hi,I follow this project,but have a problem when esp32 power on,it always restart not to working,
what can I take it and next step ? thks..

this serial port info:

ets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0010,len:4
load:0x3fff0014,len:708
load:0x40078000,len:0
load:0x40078000,len:11460
entry 0x400789f4

Brownout detector was triggered

ets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0010,len:4
load:0x3fff0014,len:708
load:0x40078000,len:0
load:0x40078000,len:11460
entry 0x400789f4

Brownout detector was triggered

ets Jun  8 2016 00:22:57

**Comment by me-no-dev** [comment]
@haosheng-huang use good usb cable and stable power supply. The error means that the supplied voltage dropped too low. Could be a bad LDO, since you did not say what board exactly you are using.

**Comment by haosheng-huang** [comment]
@me-no-dev thks,But I already use power supply only. I use this board.

![1e038497-7f1f-479e-ad52-0dbacba01c97](https://user-images.githubusercontent.com/32087651/30585941-f3020528-9d60-11e7-865b-c2129f560999.jpg)

**Comment by Spritetm** [comment]
@haosheng-huang Can I ask you to start a new issue (or post on the esp32.com forum) for this? It has nothing to do with the BLE/Arduino discussion in this issue.

**Comment by alexkirillov** [comment]
@MarosValter ,
Exactly the same "abort() was called at PC 0x400d880b on core 0" happens with my board.
After calling   BLEServer *pServer = new BLEServer();  from sample.
If I would comment everything out and try enabling code line-by-line:
ble.init("MyESP32");  - works fine
call to new BLEServer(); - makes the board to panic

@nkolban 
latest Arduino, latest ESP32_BLE.zip built from github
Board is "WeMOS" Wifi+Bluetooth+Battery (with wroom-32)

**Comment by nkolban** [comment]
@alexkirillov 
For problems, issues, enhancements and other code related items on the BLE C++ classes for Arduino as delivered by the "cpp_utils" project, can I ask that you raise an issue here:

https://github.com/nkolban/esp32-snippets

This thread/issue is about "What should we do for BLE support for ESP32 Arduino" and not so much defect hunting/chasing on a given implementation.  Currently the options for BLE support for ESP32 Arduino" are:

1. Continue with the BLE C++ classes for Arduino as delivered by cpp_utils with existing API.
2. Continue with the BLE C++ classes for Arduino as delivered by cpp_utils with alternate API.
3. Create another implementation of BLE C++ classes based on ESP-IDF as an alternate for the delivered by cpp_utils.
4. Something else ...

That thinking is still on-going and for the time being (1) exists and is gaining more users who are using it (and breaking it) in interesting ways.  This has possibly taken the pressure off making a "formal" decision but one should assume that as time passes ... a community broad decision will likely be made.

**Comment by vanilla-thunder** [comment]
Imho, BLE API should be easy to use and understandable for newbies, so people could actually concentrate on building awesome IOT devices instead of researching and studying GATT and GAP services. 
I might be a bit spoiled by JavaScript or PHP, but it seems wrong to me, if you need to copy-paste hundred(s) of cryptic code lines just to get a "hello world" program running. 
After comparing example code for CurieBLE, BLEperipheral and mbed BLE, my favorite is CurieBLE.

**Comment by d3cline** [comment]
I did a git pull tonight which broke my build.
I am sure it is something obvious I am missing,
```
/home/d3cline/Arduino/libraries/BLE/src/BLEClient.cpp: In member function 'void BLEClient::gattClientEventHandler(esp_gattc_cb_event_t, esp_gatt_if_t, esp_ble_gattc_cb_param_t*)':
/home/d3cline/Arduino/libraries/BLE/src/BLEClient.cpp:166:55: error: no matching function for call to 'BLEUUID::BLEUUID(esp_gatt_id_t&)'
    BLEUUID uuid = BLEUUID(evtParam->search_res.srvc_id);
                                                       ^
In file included from /home/d3cline/Arduino/libraries/BLE/src/BLERemoteService.h:17:0,
                 from /home/d3cline/Arduino/libraries/BLE/src/BLEClient.h:18,
                 from /home/d3cline/Arduino/libraries/BLE/src/BLEClient.cpp:14:
/home/d3cline/Arduino/libraries/BLE/src/BLEUUID.h:26:2: note: candidate: BLEUUID::BLEUUID()
  BLEUUID();
  ^
/home/d3cline/Arduino/libraries/BLE/src/BLEUUID.h:26:2: note:   candidate expects 0 arguments, 1 provided
/home/d3cline/Arduino/libraries/BLE/src/BLEUUID.h:25:2: note: candidate: BLEUUID::BLEUUID(esp_gatt_srvc_id_t)
  BLEUUID(esp_gatt_srvc_id_t srcvId);
  ^
/home/d3cline/Arduino/libraries/BLE/src/BLEUUID.h:25:2: note:   no known conversion for argument 1 from 'esp_gatt_id_t' to 'esp_gatt_srvc_id_t'
/home/d3cline/Arduino/libraries/BLE/src/BLEUUID.h:24:2: note: candidate: BLEUUID::BLEUUID(uint8_t*, size_t, bool)
  BLEUUID(uint8_t* pData, size_t size, bool msbFirst);
  ^
/home/d3cline/Arduino/libraries/BLE/src/BLEUUID.h:24:2: note:   candidate expects 3 arguments, 1 provided
/home/d3cline/Arduino/libraries/BLE/src/BLEUUID.h:23:2: note: candidate: BLEUUID::BLEUUID(esp_bt_uuid_t)
  BLEUUID(esp_bt_uuid_t uuid);
  ^
/home/d3cline/Arduino/libraries/BLE/src/BLEUUID.h:23:2: note:   no known conversion for argument 1 from 'esp_gatt_id_t' to 'esp_bt_uuid_t'
/home/d3cline/Arduino/libraries/BLE/src/BLEUUID.h:22:2: note: candidate: BLEUUID::BLEUUID(uint32_t)
  BLEUUID(uint32_t uuid);
  ^
/home/d3cline/Arduino/libraries/BLE/src/BLEUUID.h:22:2: note:   no known conversion for argument 1 from 'esp_gatt_id_t' to 'uint32_t {aka unsigned int}'
/home/d3cline/Arduino/libraries/BLE/src/BLEUUID.h:21:2: note: candidate: BLEUUID::BLEUUID(uint16_t)
  BLEUUID(uint16_t uuid);
  ^
/home/d3cline/Arduino/libraries/BLE/src/BLEUUID.h:21:2: note:   no known conversion for argument 1 from 'esp_gatt_id_t' to 'uint16_t {aka short unsigned int}'
/home/d3cline/Arduino/libraries/BLE/src/BLEUUID.h:20:2: note: candidate: BLEUUID::BLEUUID(std::__cxx11::string)
  BLEUUID(std::string uuid);
  ^
/home/d3cline/Arduino/libraries/BLE/src/BLEUUID.h:20:2: note:   no known conversion for argument 1 from 'esp_gatt_id_t' to 'std::__cxx11::string {aka std::__cxx11::basic_string<char>}'
/home/d3cline/Arduino/libraries/BLE/src/BLEUUID.h:18:7: note: candidate: constexpr BLEUUID::BLEUUID(const BLEUUID&)
 class BLEUUID {
       ^
/home/d3cline/Arduino/libraries/BLE/src/BLEUUID.h:18:7: note:   no known conversion for argument 1 from 'esp_gatt_id_t' to 'const BLEUUID&'
/home/d3cline/Arduino/libraries/BLE/src/BLEUUID.h:18:7: note: candidate: constexpr BLEUUID::BLEUUID(BLEUUID&&)
/home/d3cline/Arduino/libraries/BLE/src/BLEUUID.h:18:7: note:   no known conversion for argument 1 from 'esp_gatt_id_t' to 'BLEUUID&&'
/home/d3cline/Arduino/libraries/BLE/src/BLEClient.cpp:167:94: error: no matching function for call to 'BLERemoteService::BLERemoteService(esp_gatt_id_t&, BLEClient*)'
    BLERemoteService* pRemoteService = new BLERemoteService(evtParam->search_res.srvc_id, this);
                                                                                              ^
In file included from /home/d3cline/Arduino/libraries/BLE/src/BLEClient.h:18:0,
                 from /home/d3cline/Arduino/libraries/BLE/src/BLEClient.cpp:14:
/home/d3cline/Arduino/libraries/BLE/src/BLERemoteService.h:29:2: note: candidate: BLERemoteService::BLERemoteService(esp_gatt_srvc_id_t, BLEClient*)
  BLERemoteService(esp_gatt_srvc_id_t srvcId, BLEClient* pClient);
  ^
/home/d3cline/Arduino/libraries/BLE/src/BLERemoteService.h:29:2: note:   no known conversion for argument 1 from 'esp_gatt_id_t' to 'esp_gatt_srvc_id_t'
/home/d3cline/Arduino/libraries/BLE/src/BLERemoteService.h:27:7: note: candidate: BLERemoteService::BLERemoteService(const BLERemoteService&)
 class BLERemoteService {
       ^
/home/d3cline/Arduino/libraries/BLE/src/BLERemoteService.h:27:7: note:   candidate expects 1 argument, 2 provided
exit status 1
Error compiling for board ESP32 Dev Module.
```

**Comment by d3cline** [comment]
I noticed an upstream change has broke BLE we had. I still feel this should be something the chip manuf cares about if they want the chip to live. I know if I was a rival chip manuf I would do everything I could to provide a sane arduino BLE stack to kill it because its clear the demand is here. Also I have the WiMos Battery and have beaten them up, I notice it has issues with the same brownout error as people are describing if the battery is low. Try removing the battery to see if this solves it, if so use a better battery and make sure it has a decent charge.

**Comment by d3cline** [comment]
https://github.com/nkolban/esp32-snippets/issues/82 :(

**Comment by d3cline** [comment]
To be quite blunt, why don't the people who work on the IDF pay attention to customer demand? With a little communication they could have had this built in no time. You know lots of other hardware manufactures ignored what people wanted and wasted tons of time and money on boondoggles too, they all failed. If I can order some nordic chips and have them working before this is fixed I likely will never buy another espriff chip again. You may take this as rude, but it is just the nature of the beast and I am being honest.

**Comment by kriswiner** [comment]
As an exclusive Arduino IDE user I can sympathize. Yet I know that the
espressif team is beavering away exposing the rich capabilities of a
complicated MCU to both SDK and Arduino users. The espressif team have made
tremendous progress this year and they have been very responsive to
reported bugs and issues. It was just a few months ago I couldn't even use
I2C with the ESP32. I have no complaints, quality takes time and the
espressif team is small amd busy with a lot of issues to work out.  Both
sets of users just have to wait, and the Arduino users are last in line,
just a fact of life. This will be a wonderful, capable choice for embedded
Arduino applications requiring BLE in about six months.

In the meantime, I am using the nRF52 with Sandeep Mistry's Arduino core
<https://github.com/sandeepmistry/arduino-nRF5> and the BLEPeripheral
<https://github.com/sandeepmistry/arduino-BLEPeripheral> API to do just
about everything I need. There is no similar BLECentral API yet, sorely
missed, but for many connected Arduino applications a board like this
<https://www.tindie.com/products/onehorse/nrf52832-development-board/?pt=full_prod_search>
and some sample sketches like these
<https://github.com/kriswiner/nRF52832DevBoard> should suffice.
<https://www.tindie.com/products/onehorse/nrf52832-development-board/?pt=full_prod_search>

**Comment by DonWasabi** [comment]
I have mixed feelings about this, on the one hand: The few devs expressif pays to build the software seem to do a great job and the esp32 is one hell of a value for money beast which works pretty well on arduino, except BLE of course. As i heard the IDF 3.x is on the rise (my IDE of choice platformio still uses IDF 2.2x for arduino programming) so there might be some changes to the new IDF BLE part which may have delayed the arduino BLE work until the new IDF BLE APIs are stable? Just a guess from my side... 

On the other hand: Well.. BLE was one of the main features why i decided to use this uc in my projects. My workaround is to use external components for BLE support until the arduino BLE API is ready. That may not be a perfect solution but at least i can work on other parts in the meantime while using the same communication interface.

What i would really appreciate is some kind of feedback from the BLE arduino devs, just to see how work is going on... is the work on hold? Is development done while we are speaking? Or do you even have some kind of release date in mind? If work is on hold or delayed, what are the reasons? Are there some parts where we could even help?

**Comment by d3cline** [comment]
The issue is explained by nkolban at https://github.com/nkolban/esp32-snippets/issues/121#issuecomment-335476433 Although I understand 'things take time' things in IT die quick, and there are always new solutions popping up, or cheaper old solutions, rivals trying to beat you to the punch (which if you are an independant app dev is the worst). Having something break like this is also frustrating. Why not base arduino on a stable release? When will this IDF development which breaks everything stop? ESP32 was supposed to be 'the 8266 + BLE' otherwise all the great things it can do can mostly be done by the 8266 (touch and audio is alright, but not even close to BLE). I would have never known or cared who espressif was if the 8266 wasn't so arduino friendly. So far it isn't that, and based on the track record for this issue it won't ever be that by the time it matters. And I think that is what upsets me, realizing this chip has a lot of potential which may be ignored because it will be obsolete by the time it is done, and all this time and effort I spent waiting for this solution when I should have just gave up on it months ago and used a nordic working solution. That this could be easily fixed if there was some kind of communication or priority to make BLE stable and functioning with a simple API. The functionality is there, it's just a pain to get to, I tend to feel the people who wrote those functions could easily write some wrappers but its 'community software' so why bother. Just because arduino is a community project should not mean it has the last priority. The licensing from my point of view is just as commercial as their IDF, more since I can rapidly develop things as a lone developer. The bottom line is if they care about selling chips and maintaining customer trust they might want to fix this and stop breaking it, make BLE a priority or expect we will move on.

**Comment by kriswiner** [comment]
"the people who wrote those functions could easily write"

It is never easy.

And Nordic is no paragon of virtue. Their I2S has an error baked into the
silicon. Their wire solution is a joke. Their SDK is the worst kind of
spaghetti code. The Arduino core based on the SDK is barely adquate.

And don't get me started on ST, whose HAL is full of halfbaked "solutions"
that might work for thier demos but not in real applications.

I am a hardware guy but even I know writing good software ain't easy and it
is very time consuming to get it right.

I don't think whining and moaning is helping anyone. Maybe you should learn
to use the SDK, at least this is more stable and more capable today.

**Comment by edcasati** [comment]
I was just hoping to be able to receive simple serial characters on an ESP32 using Arduino by now. Basically an HC-06 type functionality to enable some remote control.
________________________________

**Comment by d3cline** [comment]
I already ordered a RedBear and Adafruit AT+nordic board, should be just fine and only marginally more expensive if I order them from China. And will spend my time making them function with the examples I already have so I can move onto the mobile app and all the other stuff I have to concern myself with. I will likely follow this thread but don't hold much hope this will be done by the time it matters. I am not whining and moaning, I am expressing my point of view. It is an industry, with a market and all the other economic facts which go with that, and those things matter if you are here to make products and not projects.

**Comment by kriswiner** [comment]
For this I would recommend an nRF52
<https://www.tindie.com/products/onehorse/nrf52-add-on-for-butterfly-and-teensy/?pt=full_prod_search>
using the NUS for now.

**Comment by kriswiner** [comment]
ESP32 is not ready for products yet.

**Comment by ibaranov-cp** [comment]
Bit of a conflict of interest there @kriswiner , recommending your own products.. ;)

**Comment by kriswiner** [comment]
Hardly, I also make ESP8285 and ESP32 products and when appropriate I would
recommend them also. Anyone expecting to be able to use the ESP32 in a
consumer product will be sadly disappointed today; I hope this is not the
case six months from now. In the meantime, recommending another solution
isn't a conflict of interest. My products are the handiest example. But my
guess is anyone developing consumer products will not use a develpment
board from me or anyone else.

**Comment by me-no-dev** [comment]
@kriswiner I could agree that sticking a competitor’s product in our face here is the least not nice ;) but whatever :D BTW @nkolban ’s BLE is now included ;)

**Comment by kriswiner** [comment]
No malice intended;>

Your competition is far from perfect. It's just that commercial designers
need to use products that work. Up to today (glad to hear BLE is now
available) BLE on the ESP32 didn't.

**Comment by Nakul93** [comment]
How to increase the stack size I am using Wifi + BLE but when I use the wifimanager library it crashes. So I have to either comment wifi setup or ble setup. Please help

**Comment by MartyMacGyver** [comment]
@Nakul93 I'd think adjusting the stack values in `hardware/espressif/esp32/tools/sdk/include/config/sdkconfig.h` might help. `bt.h` seems to use that and `hardware/espressif/esp32/tools/sdk/include/esp32/esp_task.h` as a basis for setting up their stacks. I'm not sure which stack value might be causing you problems though.

**Comment by haarislabs** [comment]
Is it still the same case ?? My team is planning to use Esp32 for a biomedical product in which reliability is a major factor, so @kriswiner  do you still think esp32 is not ready for consumer products?
As of now I was testing all the features of esp32 (wifi,ble) using idf sdk and testing if everything would work for a quick prototype.

**Comment by kriswiner** [comment]
" As of now I was testing all the features of esp32 (wifi,ble) using idf
sdk and testing if everything would work for a quick prototype. "

If you tested all of the ESP32 features you need for your application and
they all work well for you then yes, this would be a good choice. In fact,
I would recommend the ESP32 Pico D4. Very easy to use.

I use it myself, see here
<https://hackaday.io/project/71283-small-esp32-module-for-easier-custom-design>
.

**Comment by beegee-tokyo** [comment]
Same as @kriswiner , I tried to test all features, but using Arduino framework. Most of the stuff is working reliable now. Wrote some blogs about it [ESP32](https://desire.giesecke.tk/index.php/esp32/)

Working a customer project right now where we will use I2C to poll a sensor and send the data over BLE  every 5ms. But because it (hopefully) will be a commercial product, we do not use Arduino framework for it, instead I program everything using directly ESP-IDF. First prototype code is working now. (Surprised that we could achieve the 5ms update rate over BLE).

**Comment by aster94** [comment]
I am not sure if you already made a decision about which API to use
if not: arduino's official API (curieBLE) 👍
if yes: please update the first post to prevent comment like this 😅

**Comment by atanisoft** [comment]
@me-no-dev I think this one can be closed since there is a BLE lib in arduino-esp32 already
